sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -246,6 +247,54 @@ class Analyzer(
       CleanupAliases)
   )
 
+  /**
+   * 1. Turns Add/Subtract of DateType/TimestampType/StringType and CalendarIntervalType
+   *    to TimeAdd/TimeSub.
+   * 2. Turns Add/Subtract of TimestampType/DateType/IntegerType
+   *    and TimestampType/IntegerType/DateType to DateAdd/DateSub/SubtractDates and
+   *    to SubtractTimestamps.
+   * 3. Turns Multiply/Divide of CalendarIntervalType and NumericType
+   *    to MultiplyInterval/DivideInterval
+   */
+  case class ResolveBinaryArithmetic(conf: SQLConf) extends Rule[LogicalPlan] {
+    override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsUp {
+      case p: LogicalPlan => p.transformExpressionsUp {
+        case UnresolvedAdd(l, r) => (l.dataType, r.dataType) match {
+          case (TimestampType | DateType | StringType, CalendarIntervalType) =>
+            Cast(TimeAdd(l, r), l.dataType)
+          case (CalendarIntervalType, TimestampType | DateType | StringType) =>
+            Cast(TimeAdd(r, l), r.dataType)
+          case (DateType, _) => DateAdd(l, r)
```

Comment:
Hi @cloud-fan, when I was implementing this, I got a bit confused.
Our date_sub/date_add actually support implicit type cast, that is
```
select date_add(date'1900-01-01', 1);
select date_add(date'1900-01-01', 1.2);
select date_add(date'1900-01-01', "1.2");
```
are supported, and result the same.

but while using `date'1900-01-01' + 1`, we only support int. Is this nonconforming？

Reply:
This part is really messy. @marope AFAIK there is a proposal to update the type coercion rules to follow SQL standard, is it still active?

Reply:
Em.. I guess that you are pinging @maropu?

Reply:
Ur, I think there is currently no active work about that... cc: @gengliangwang 
@yaooqinn Cannot we easily prohibit the implicit cast for date adds? That's confusing regardless of the standard.

Reply:
From hive
```
DATE_ADD() only takes TINYINT/SMALLINT/INT types as second argument, got DOUBLE
```

Reply:
@maropu It's true that there no active work about that. We should revisit and try creating a full plan next Q1/Q2.